Content.Shared/Doors/Systems/SharedDoorSystem.Airlock.cs

```diff
@@ -0,0 +1,189 @@
+using Content.Shared.Doors.Components;
+using Content.Shared.Prying.Components;
+using Content.Shared.Wires;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.Doors.Systems;
+
+public abstract partial class SharedDoorSystem
+{
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] protected readonly SharedDoorSystem DoorSystem = default!;
+    [Dependency] private readonly SharedWiresSystem _wiresSystem = default!;
+
+    private void InitializeAirlock()
+    {
+        SubscribeLocalEvent<AirlockComponent, BeforeDoorClosedEvent>(OnBeforeDoorClosed);
+        SubscribeLocalEvent<AirlockComponent, DoorStateChangedEvent>(OnStateChanged);
+        SubscribeLocalEvent<AirlockComponent, DoorBoltsChangedEvent>(OnBoltsChanged);
+        SubscribeLocalEvent<AirlockComponent, BeforeDoorOpenedEvent>(OnBeforeDoorOpened);
+        SubscribeLocalEvent<AirlockComponent, BeforeDoorDeniedEvent>(OnBeforeDoorDenied);
+        SubscribeLocalEvent<AirlockComponent, GetPryTimeModifierEvent>(OnGetPryMod);
+        SubscribeLocalEvent<AirlockComponent, BeforePryEvent>(OnBeforePry);
+    }
+
+    private void OnBeforeDoorClosed(Entity<AirlockComponent> airlock, ref BeforeDoorClosedEvent args)
+    {
+        if (args.Cancelled)
+            return;
+
+        if (!airlock.Comp.Safety)
+            args.PerformCollisionCheck = false;
+
+        // only block based on bolts / power status when initially closing the door, not when its already
+        // mid-transition. Particularly relevant for when the door was pried-closed with a crowbar, which bypasses
+        // the initial power-check.
+
+        if (!TryComp(airlock, out DoorComponent? door)
+            || door.State is DoorState.Opening or DoorState.Closing
+            || CanChangeState(airlock,
+                door.State is DoorState.AttemptingCloseByPrying or DoorState.AttemptingOpenByPrying))
+            return;
+
+        args.Cancel();
```

Comment:
```suggestion
        if (TryComp(airlock, out DoorComponent? door)
            && door.State is not (DoorState.Opening or DoorState.Closing)
            && !CanChangeState(airlock,
                door.State is DoorState.AttemptingCloseByPrying or DoorState.AttemptingOpenByPrying))
            args.Cancel();
```
imo it's kind of confusing that the "unguarded behavior" of the handler is to cancel the event.